processor/batchprocessor/batch_processor.go

```diff
@@ -93,11 +98,34 @@ type shard struct {
 	timer *time.Timer
 
 	// newItem is used to receive data items from producers.
-	newItem chan any
+	newItem chan dataItem
 
 	// batch is an in-flight data item containing one of the
 	// underlying data types.
 	batch batch
+
+	// pending describes the contributors to the current batch.
+	pending []pendingItem
+
+	// totalSent counts the number of items processed by the
+	// shard in its lifetime.
+	totalSent uint64
+}
+
+// pendingItem is stored parallel to a pending batch and records how
+// many items the waiter submitted, used to match trace contexts with
+// batches.  It is also used inside sendItems() to represent a
+// partially complete batch.
+type pendingItem struct {
+	parentCtx context.Context
```

Comment:
In a future PR, I will need the context because this function returns an { error, count } to the waiter but has to check for a canceled context, in which case the waiter has gone.

Reply:
Here is how the context is used in the downstream fork:
https://github.com/open-telemetry/otel-arrow/blob/7cceec5e8ca0c0a0ee8be4806ced19e49e96778b/collector/processor/concurrentbatchprocessor/batch_processor.go#L437